// atomicms/index.js
'use strict';

var fs = require('fs');
var root = require('app-root-path');
var Mark = require('markup-js');

var templates = {};
var config = {
  key: 'key.json',
  content: '/content',
  template: '/templates',
  organism: '/organisms'
};

var cms = function() {
  return function(req, res, next) {
    var targetTemplate = templates[req.url];
    if(targetTemplate) {
      res.send(targetTemplate.html);
    } else {
      next();
    }
  }
}

cms.init = function(settings) {
  config = settings;
  fs.readFile(root + '/' + config.key, 'utf8', function(err, file) {
    if(err) throw err;

    var data = JSON.parse(file);
    for(var key in data) {
      renderFile(data[key].template, data[key].content, function(data) {
        templates[key] = { html: data };
      });
    }
  });
}

function renderFile(filename, keyname, callback) {
  fs.readFile(root + config.template + '/' + filename, 'utf8', function(err, file) {
    if(err) throw err;

    var template = file;
    fs.readFile(root + config.content + '/' + keyname, 'utf8', function(err, file) {
      if(err) throw err;

      var content = JSON.parse(file);
      var organisms = template.match(/{{[A-Za-z\s]+}}/g);
      organisms = organisms.map(function(o) {
        var matches = o.match(/{{([A-Za-z]+)\s+as\s+([A-Za-z]+)}}/);
        return {
          full: matches[0],
          organism: matches[1],
          variable: matches[2]
        };
      });

      var i = 0;
      organisms.forEach(function(o) {
        markupOrganism(o, function(data) {
          o.markupped = data;
          i++;
          if(i === organisms.length) {
            markupTemplate(template, organisms, function(data) {
              callback(data);
            });
          }
        });
      });

      function markupOrganism(o, cb) {
        fs.readFile(root + config.organism + '/' + o.organism + '.html', 'utf8', function(err, file) {
          if(err) throw err;

          cb(Mark.up(file, content[o.variable]));
        });
      }

      function markupTemplate(template, organisms, cb) {
        var markupped = template;
        organisms.forEach(function(o) {
          markupped = markupped.replace(new RegExp(o.full), o.markupped);
        });
        cb(markupped);
      }
    });
  })
}
module.exports = cms;
